numba/jitclass/immutable.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import absolute_import, print_function
+from numba import types
+from numba.typing import templates
+from numba.datamodel import default_manager, models
+from numba.targets import imputils
+from numba import cgutils
+from .base import ClassBuilder
+
+
+class ImmInstanceModel(models.StructModel):
+    def __init__(self, dmm, fe_typ):
+        members = list(fe_typ.struct.items())
+        super(ImmInstanceModel, self).__init__(dmm, fe_typ, members)
+
+
+class ImmRefModel(models.PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        self._pointee_model = dmm.lookup(fe_type.instance_type)
+        self._pointee_be_type = self._pointee_model.get_data_type()
+        be_type = self._pointee_be_type.as_pointer()
+        super(ImmRefModel, self).__init__(dmm, fe_type, be_type)
+
+
+default_manager.register(types.ImmutableClassInstanceType, ImmInstanceModel)
+default_manager.register(types.ImmutableClassType, models.OpaqueModel)
+default_manager.register(types.ImmutableClassRefType, ImmRefModel)
+
+
+class ImmutableClassBuilder(ClassBuilder):
+    instance_type_class = types.ImmutableClassInstanceType
+
+    def implement_frontend(self, instance_type):
+        self.implement_attribute_typing(instance_type)
+        self.register_byref_attribute_typing(instance_type)
+
+    def register_byref_attribute_typing(self, instance_type):
+        class RefStructAttribute(templates.AttributeTemplate):
```

Comment:
If this is the same template for all reference types, why not define it at the top level with `ImmutableClassRefType` as the key? Or am I missing something?